src/Microsoft.ML.Dnn/DnnRetrainTransform.cs

```diff
@@ -811,18 +812,20 @@ private sealed class Mapper : MapperBase
                             throw Contracts.Except($"Input shape mismatch: Input '{_parent._inputs[i]}' has shape {originalShape.ToString()}, but input data is of length {typeValueCount}.");
 
                         // Fill in the unknown dimensions.
-                        var l = new int[originalShape.NDim];
-                        for (int ishape = 0; ishape < originalShape.NDim; ishape++)
-                            l[ishape] = originalShape[ishape] == -1 ? (int)d : originalShape[ishape];
+                        var originalShapeDims = originalShape.dims;
+                        var originalShapeNdim = originalShape.ndim;
+                        var l = new int[originalShapeNdim];
+                        for (int ishape = 0; ishape < originalShapeNdim; ishape++)
+                            l[ishape] = originalShapeDims[ishape] == -1 ? (int)d : originalShapeDims[ishape];
```

Comment:
>(int)d [](start = 74, length = 6)

Should this cast also be done outside the loop?

Reply:
This is not an addition in this PR. If required, I can do this change

---
In reply to: [324450404](https://github.com/dotnet/machinelearning/pull/4205#discussion_r324450404) [](ancestors = 324450404)